spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/SkipperAppRegistryController.java

```diff
@@ -230,9 +249,58 @@ public void makeDefault(@PathVariable("type") ApplicationType type, @PathVariabl
 	@ResponseStatus(HttpStatus.OK)
 	public void unregister(@PathVariable("type") ApplicationType type, @PathVariable("name") String name,
 			@PathVariable("version") String version) {
+
+		// AppRegistration defaultApp = appRegistryService.getDefaultApp(name, type);
+		if (type != ApplicationType.task) { // && defaultApp != null && defaultApp.getVersion().equals(version)) {
+			String streamWithApp = findStreamContainingAppOf(type, name, version);
+			if (streamWithApp != null) {
+				throw new AppUsedInStreamException(String.format("The app [%s:%s:%s] you're trying to unregister is " +
+						"already in use on %s stream", name, type, version, streamWithApp));
+			}
+		}
+
 		appRegistryService.delete(name, type, version);
 	}
 
+	private String findStreamContainingAppOf(ApplicationType appType, String appName, String appVersion) {
+		Iterable<StreamDefinition> streams = streamDefinitionRepository.findAll();
+		for (StreamDefinition stream : streams) {
+			StreamDeployment streamDeployment = this.streamService.info(stream.getName());
+			for (StreamAppDefinition streamAppDefinition : stream.getAppDefinitions()) {
+				final String streamAppName = streamAppDefinition.getRegisteredAppName();
+				final ApplicationType streamAppType;
+				try {
+					streamAppType = DataFlowServerUtil.determineApplicationType(streamAppDefinition);
+				}
+				catch (CannotDetermineApplicationTypeException e) {
+					logger.warn("Can not determine ApplicationType for " + streamAppDefinition);
+					continue;
+				}
+				if (appType != streamAppType) {
+					break;
+				}
+				Map<String, Map<String, String>> streamDeploymentPropertiesMap = new HashMap<>();
+				String streamDeploymentPropertiesString = streamDeployment.getDeploymentProperties();
+				ObjectMapper objectMapper = new ObjectMapper();
+				try {
+					streamDeploymentPropertiesMap = objectMapper.readValue(streamDeploymentPropertiesString,
+							new TypeReference<Map<String, Map<String, String>>>() {
+							});
+				}
+				catch (IOException e) {
+					throw new RuntimeException("Can not deserialize string " + streamDeploymentPropertiesString);
+				}
+				if (streamDeploymentPropertiesMap.containsKey(appName)) {
+					Map<String, String> appDeploymentProperties = streamDeploymentPropertiesMap.get(streamAppName);
+					if (appDeploymentProperties.containsValue(appVersion)) {
```

Comment:
Looking in only values is because they key to find the version is the resource string.  As an example of the current deployment properties format

```
Stream Deployment properties: {
  "log" : {
    "spring.cloud.deployer.group" : "httptest",
    "maven://org.springframework.cloud.stream.app:log-sink-rabbit" : "1.1.0.RELEASE"
  },
  "http" : {
    "spring.cloud.deployer.group" : "httptest",
    "maven://org.springframework.cloud.stream.app:http-source-rabbit" : "1.2.0.RELEASE"
  }
}
```

This should instead be 

```
Stream Deployment properties: {
  "log" : {
    "spring.cloud.deployer.group" : "httptest",
    "resource" :  "maven://org.springframework.cloud.stream.app:log-sink-rabbit", 
    "version" : "1.1.0.RELEASE"
  },
  "http" : {
    "spring.cloud.deployer.group" : "httptest",
    "resource" : "maven://org.springframework.cloud.stream.app:http-source-rabbit",
    "version" : "1.2.0.RELEASE"
  }
}
```